### PR TITLE
79 bulk delete async

### PR DESCRIPTION
**Pull request by newtewt**
# Description

Bulk delete changes for asyn/await and angular 10

https://github.com/medic/angular10-migration/issues/79

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/e2e/bulk-delete.js [hunk 1]
```diff
@@ -56,35 +56,29 @@ describe('Bulk delete reports', () => {
...
-        results.forEach(result => {
-          savedUuids.push(result.id);
-        });
-        done();
-      })
-      .catch(done.fail);
+  beforeEach(async () => {
+    const results = await Promise.all(docs.map(utils.saveDocNative));
+    results.forEach(result => {
+      savedUuids.push(result.id);
```
Comment (craig-landry) [inline tests/e2e/bulk-delete.js:62]: When I first read this I was going to point out the awkward appending here instead of assignment. Then I realized it is on purpose (at least if the original code was "on purpose"). This shared state risks making tests dependent on each other having run (something we'd all talked about recently as wanting to get away from). I don't think today is the day to correct this, but this is a good one to remember for future improvements.

For now though, thanks for the async/await treatment here. This is much easier to read now!

File: tests/e2e/bulk-delete.js [hunk 2]
```diff
@@ -56,35 +56,29 @@ describe('Bulk delete reports', () => {
...
+    });
   });
 
-  afterEach(utils.afterEach);
+  afterEach(async () => await utils.afterEachNative());
```
Comment (craig-landry) [inline tests/e2e/bulk-delete.js:66]: The `await` is just going to wait for a result, then put that result into a new pre-resolved promise and then wait on that. It is preferred that this just return the original promise with no extra one in the mix.
```suggestion
  afterEach(async () => utils.afterEachNative());
```

File: tests/e2e/login/db-sync-filter.spec.js [hunk 3]
```diff
@@ -189,7 +189,7 @@ describe('db-sync-filter', () => {
   };
 
   const getServerRevs = async (docIds) => {
-    const result = utils.requestOnMedicDbNative({ path: '/_all_docs', params: { keys: docIds } });
+    const result = await utils.requestOnMedicDbNative({ path: '/_all_docs', params: { keys: docIds } });
     return result.rows;
```
Comment (craig-landry) [inline tests/e2e/login/db-sync-filter.spec.js:193]: This is a good case to do the await in the function and let the async function wrap the result. In many cases we just want to do something and return the promise directly, but here the function is only intended to return `rows` and not make the caller deal with (or have to know about) any sort of `result` object in the mix. That's good design. Doing the async await like this retains that design. Good stuff!

File: tests/e2e/transitions/sms_workflows.spec.js [hunk 4]
```diff
@@ -115,7 +115,7 @@ describe('SMS workflows', () => {
   
   afterAll(async () => await utils.revertDbNative());
 
-  afterEach(async () => utils.revertDbNative(contacts.map(c => c._id)));
+  afterEach(async () => await utils.revertDbNative(contacts.map(c => c._id)));
```
Comment (craig-landry) [inline tests/e2e/transitions/sms_workflows.spec.js:118]: I was going to suggest this:
```suggestion
  afterEach(async () => utils.revertDbNative(contacts.map(c => c._id)));
```

...but I see that's what it was before. Why the change?

File: tests/page-objects/reports/reports.po.js [hunk 5]
```diff
@@ -82,30 +91,70 @@ module.exports = {
...
   },
 
+  deleteSelectedReportsNative: async (savedUuids) => {
+    await element(by.css('.action-container .detail-actions .delete-all')).click();
+    const confirmButton = element(by.css('.btn.submit.btn-danger'));
+    await helper.waitElementToBeClickable(confirmButton, 5000);
+    await confirmButton.click();
+    await helper.waitElementToBeClickable(confirmButton, 5000);
+    await confirmButton.click();
+    await helper.waitElementToPresent(element(by.css('#reports-list li')), 30000);
```
Comment (craig-landry) [inline tests/page-objects/reports/reports.po.js:101]: Yikes, are there really cases where this can take 30 seconds!?

File: tests/page-objects/reports/reports.po.js [hunk 6]
```diff
@@ -82,30 +91,70 @@ module.exports = {
...
+    utils.deprecated('deselectAll','deselectAllNative');
     element(by.css('.action-container .deselect-all')).click();
     expect(element(by.css('#reports-content .selection-count > span:first-child')).isDisplayed()).toBeFalsy();
     expect(element.all(by.css(reportBody)).count()).toBe(0);
   },
 
+  deselectAllNative: async () => {
+    await element(by.css('.action-container .deselect-all')).click();
+    expect(await element(by.css('#reports-content .selection-count > span')).isPresent()).toBeFalsy();
+    expect(await element.all(by.css(reportBody)).count()).toBe(0);
```
Comment (craig-landry) [inline tests/page-objects/reports/reports.po.js:117]: These `expect` checks are effectively "test logic" and seeing them here in a Page Object file is unexpected (no pun intended). Right now the focus is on getting things migrated, so it's ok to keep it as-is, but over time let's work to keep all the `expect` calls in the tests themselves and keep Page Objects as helpers for just doing the basic reading of things on the page.

File: tests/page-objects/reports/reports.po.js [hunk 7]
```diff
@@ -136,5 +209,12 @@ module.exports = {
       'There should be at least one report in the LHS'
     );
   },
+
+  waitForReportToAppearNative: async (numOfReports = 1) => {
+    await browser.refresh();
+    while (await module.exports.allReports().count < numOfReports) {
+      browser.sleep(100);
```
Comment (craig-landry) [inline tests/page-objects/reports/reports.po.js:216]: Without the `await`, this line won't have any "sleep" impact.
```suggestion
      await browser.sleep(100);
```

**Review by craig-landry — CHANGES_REQUESTED** [review]

File: tests/e2e/login/db-sync-filter.spec.js (hunk 3 above)
Comment (dianabarsan) [reply tests/e2e/login/db-sync-filter.spec.js:193]: whops!

File: tests/e2e/transitions/sms_workflows.spec.js (hunk 4 above)
Comment (newtewt) [reply tests/e2e/transitions/sms_workflows.spec.js:118]: The eslint plugin we added fails when you leave the await out.  

```
Running "exec:eslint" (exec) task
Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db

/home/newt/dev/cht-core/tests/e2e/transitions/sms_workflows.spec.js
  118:13  warning  Missing await in async function  async/missing-await-in-async-fn

✖ 1 problem (0 errors, 1 warning)
```

Comment (dianabarsan) [reply tests/e2e/transitions/sms_workflows.spec.js:118]: Can we change it so it's an error instead of a warning?

File: tests/page-objects/reports/reports.po.js (hunk 5 above)
Comment (newtewt) [reply tests/page-objects/reports/reports.po.js:101]: I didn't even see that was 30 seconds. I'm going to remove that and let the default 10 seconds.

File: tests/e2e/transitions/sms_workflows.spec.js (hunk 4 above)
Comment (dianabarsan) [reply tests/e2e/transitions/sms_workflows.spec.js:118]: My point is... most of the time, people won't scroll up to the top of their test output to see if they had lint warnings. So most of the time, they'll go unnoticed.

Comment (newtewt) [reply tests/e2e/transitions/sms_workflows.spec.js:118]: Yeah that's my fault. I didn't look at what the 1 meant and assumed the eslint warning was enough. I have switched it locally to error but I'm waiting on @craig-landry  feedback if we want to stick to the syntax he outlines vs the syntax that doesn't break the plugin rule.

Comment (craig-landry) [reply tests/e2e/transitions/sms_workflows.spec.js:118]: I agree with the preference of errors over warnings.

Regarding this one in particular.... :disappointed:  
options seem to be:
1. Leave the code as what I consider "correct" and add a matching "ignore" for the tooling. This is really annoying since those ignore statements are supposed to be more about why some case is a special exemption to an otherwise good rule.
2. Change the code to add the `await` as seen here. Making things worse (in my opinion here) due to a tools constraint is very frustrating. It also means a future reader (probably me again every time I see it) will want to "fix" it.
3. Change the function syntax to include a `return`. I am curious if that makes the tooling evaluate it differently. If it does, this is technically better than option 2, since it has no actual impact to what is happening, it's just an alternative syntax that is equally correct, but just less optimal.
4. Maybe I just have this all wrong and the tool is saving me from my ignorance. This is actually the best outcome because then we can feel comfortable with the tool and the universe is safe.

In the end here, I'm hoping for 4. If that's not it, I can calm down and let this go as 2.

File: tests/e2e/bulk-delete.js (hunk 1 above)
Comment (newtewt) [reply tests/e2e/bulk-delete.js:62]: To clarify you are referring to the `savedUuids` being shared right?

Comment (craig-landry) [reply tests/e2e/bulk-delete.js:62]: Yup.

**Comment by newtewt** [comment]
I believe I have updated everything here @craig-landry

**Review by dianabarsan — CHANGES_REQUESTED** [review]
I found a couple of merge misses :(

**Comment by newtewt** [comment]
@dianabarsan  I've updated, can you have another look?

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Just 2 tiny tiny observations! Awesome work! And thanks a lot for cleaning up the branch!

File: tests/e2e/bulk-delete.js [hunk 8]
```diff
@@ -56,35 +56,29 @@ describe('Bulk delete reports', () => {
...
-      .catch(done.fail);
+  beforeEach(async () => {
+    const results = await utils.saveDocs(docs);
+    results.forEach(result => {
+      savedUuids.push(result.id);
+    });
   });
 
-  afterEach(utils.afterEach);
+  afterEach(() => utils.afterEach);
```
[the remark this answers is not in the document]
Comment (newtewt) [reply tests/e2e/bulk-delete.js:66]: Thanks! I'm still getting my head around the different ways you can do make these calls.

**Comment by newtewt** [comment]
@dianabarsan  I've cleaned up based on your comments. Have a look again?

**Review by dianabarsan — APPROVED** [review]
Awesome work!
